enhancements/disaster-recovery/automated-cluster-backups.md

```diff
@@ -0,0 +1,324 @@
+        one-time backup.
+      - Default: `""`
+    - spec.schedule `string`
+      - [Standard cron expression](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#cron-schedule-syntax)
+      - Results in backups scheduled repeatedly to successive control plane
+        nodes.
+      - Default: `""`
+    - spec.maxDurationWithoutBackup `duration`
+      - If a backup has not successfully completed in the configured
+        interval, a critical alert will be generated for this and every
+        subsequent interval. The alert will either indicate that backups are
+        not configured, or supply the most recent reason a backup failed.
+      - Setting to `0` will disable alerting.
+      - Default: `1h`
+    - spec.retentionCount `int`
+      - The maximum number of backups to retain.
+      - If the number of succesful backups (as determined by recorded
+        status) matches `retentionCount`, the oldest backup will removed
+        before a new backup is initiated.
+      - Default: 15
+
+- TODO Need a way to track backups once initiated
+  - New API type?
+  - Metadata to store
+    - openshift version
+    - node the backup was performed on
+      - Would also be where the backup was stored
+    - path the backup was written to
+    - time
+    - size
+    - success/failure
+      - failure reason
+  - Stored as spec or status?
+  - Remove backup api resources as part of removing backups
+
+
+- Implement a new controller that responds to the backup configuration
+  - The new controller will initially be shipped with the etcd operator for simplicity
+  - If `maxDurationWithoutBackup` is non-zero and is exceeded without a
+    successfull backup, create a critical alert. This is intended to ensure
+    that cluster admins are encouraged to configure automated backups if only
+    to disable the alerting.
+  - The cost of taking automated backups should be spread over all the
+    control plane nodes to minimize the IO impact on any one node.
+    - Spreading backups across available nodes has the added benefit of
+      maximizing the chances of retaining a viable backup should one or more
+      control plane nodes be lost.
+
+- If a backup is configured to occur due to `backupReason` or `schedule`:
+  - Given the set of control plane nodes
+  - Given the set of the successful backups recorded in status
+  - Only a single backup pod should be running at a given time. If a backup
+    pod is still running:
+    - If the pod has been running longer than a maximum period (TODO How
+      long?), terminate it and record the backup as having failed.
+    - If the pod has been running for less than the maximum period, defer the
+      scheduled backup for (TODO How long)
+  - If the number of successful backups matches `retentionCount`
+    - Ensure removal of the oldest backup from the node that it was written to
+    - Remove record of removed backup from the api?
+    - Log removal of the backup
```

Comment:
log before you remove the record to make an indempotent command.

Reply:
Done